feg/gateway/services/n7_n40_proxy/notify/notify_handler.go

```diff
@@ -0,0 +1,210 @@
+	}
+	smPolicyDecision := smUpdateNotify.SmPolicyDecision
+	sessionId, imsi, err := getSessionIdAndIMSI(ctx.Param(EncodedSessionId))
+	if err != nil {
+		glog.Errorf("Unable to fetch session-id for UpdateNotify- %s", err)
+		return err
+	}
+
+	policyReauthProto := n7.GetPolicyReauthRequestProto(sessionId, imsi, smPolicyDecision)
+	ans, err := client.PolicyReAuth(context.Background(), policyReauthProto)
+	if err != nil {
+		glog.Errorf("Error relaying N7 policy reauth request to gateway: %s", err)
+		return err
+	}
+	partialSuccessReport := n7.GetPartialSuccessReportN7(ans)
+	if partialSuccessReport == nil {
+		return ctx.NoContent(http.StatusOK)
+	}
+	return ctx.JSON(http.StatusOK, partialSuccessReport)
+}
+
+func (handler *NotificationHandler) PostSmPolicyTerminateNotification(ctx echo.Context) error {
+	client, err := relay.GetAbortSessionResponderClient(handler.cloudRegistry)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	defer client.Close()
+
+	var smTermNotify n7_client.TerminationNotification
+	err = ctx.Bind(&smTermNotify)
+	if err != nil {
+		glog.Errorf("Invalid SmPolicyNotification received: %s", err)
+		return ctx.NoContent(http.StatusBadRequest)
+	}
+
+	// smPolicyUrl := string(smTermNotify.ResourceUri)
+	sessionId, imsi, err := getSessionIdAndIMSI(ctx.Param(EncodedSessionId))
+	if err != nil {
+		glog.Errorf("Unable to fetch session-id for TerminateNotify - %s", err)
+		return err
+	}
+
+	ans, err := client.AbortSession(context.Background(), &protos.AbortSessionRequest{
+		UserName:  imsi,
+		SessionId: sessionId,
+	})
+	if err != nil {
+		glog.Errorf("Error relaying ASR to gateway: %s", err)
+		return nil
+	}
+
+	switch ans.Code {
+	case protos.AbortSessionResult_SESSION_NOT_FOUND:
+		err = ctx.NoContent(http.StatusNotFound)
+	case protos.AbortSessionResult_USER_NOT_FOUND:
+		err = ctx.NoContent(http.StatusNotFound)
+	case protos.AbortSessionResult_GATEWAY_NOT_FOUND:
+		err = ctx.NoContent(http.StatusInternalServerError)
+	case protos.AbortSessionResult_RADIUS_SERVER_ERROR:
+		err = ctx.NoContent(http.StatusInternalServerError)
+	case protos.AbortSessionResult_SESSION_REMOVED:
+		err = ctx.NoContent(http.StatusNoContent)
+	default:
+		err = ctx.NoContent(http.StatusNoContent)
+	}
+	return err
+}
+
+// waitForServer waits for the Echo server to be launched and returns the Listener Address.
+func waitForServer(e *echo.Echo, errChan <-chan error) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 200*time.Millisecond)
+	defer cancel()
+
+	ticker := time.NewTicker(5 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return "", ctx.Err()
+		case <-ticker.C:
+			addr := e.ListenerAddr()
+			if addr != nil && strings.Contains(addr.String(), ":") {
+				// server started
+				return addr.String(), nil
+			}
+		case err := <-errChan:
+			if err == http.ErrServerClosed {
+				return "", nil
+			}
+			return "", err
+		}
+	}
+}
```

Comment:
I have seen this function before, maybe it would be good to make it part of the sbi client itself

Reply:
Done